pkg/client/client.go

```diff
@@ -0,0 +1,262 @@
+package client
+
+import (
+	"encoding/base64"
+	"encoding/json"
+	"errors"
+	"net"
+
+	"github.com/kevinburke/nacl"
+	"github.com/sirupsen/logrus"
+)
+
+type ServerResponse struct {
+	Action            Action          `json:"action"`
+	Message           string          `json:"message,omitempty"`
+	Nonce             string          `json:"nonce"`
+	PublicKey         string          `json:"publicKey,omitempty"`
+	Error             string          `json:"error,omitempty"`
+	ErrorCode         json.RawMessage `json:"errorCode,omitempty"`
+	DecryptedResponse []byte          `json:"-"`
+}
+
+type AssociateResponse struct {
+	Hash    string `json:"hash"`
+	ID      string `json:"id"`
+	Nonce   string `json:"nonce"`
+	Success string `json:"success"`
+	Version string `json:"version"`
+}
+
+type Entry struct {
+	Group    string `json:"Group"`
+	Login    string `json:"Login"`
+	Name     string `json:"Name"`
+	Password string `json:"password"`
+	UUID     string `json"uuid"`
+	TOTP     string `json:"totp,omitempty"`
+}
+
+type EntriesResponse struct {
+	Count   int      `json"count"`
+	Entries []*Entry `json"entries"`
+}
+
+type Client struct {
+	socket     net.Conn
+	socketPath string
+	crypt      Crypt
+	ClientID   string
+}
+
+const ClientID string = "gokeexc"
+
+func NewClient(sockPath string, assoName string, assoB64Key string) Client {
+	assoKey := nacl.NewKey()
+	if len(assoB64Key) != 0 {
+		assoKey = B64ToNaclKey(assoB64Key)
+	}
+
+	crypt := Crypt{
+		AssociatedName: assoName,
+		AssociatedKey:  assoKey,
+	}
+
+	crypt.NewKeys()
+	return Client{
+		socketPath: sockPath,
+		crypt:      crypt,
+		ClientID:   ClientID + NaclNonceToB64(nacl.NewNonce()),
+	}
+}
+
+func (c *Client) Connect() error {
+	var err error
+	logrus.Debugf("connect to socket %s", c.socketPath)
+	c.socket, err = net.DialUnix("unix", nil, &net.UnixAddr{Name: c.socketPath, Net: "unix"})
+	return err
+}
+
+func (c *Client) Disconnect() error {
+	if c.socket != nil {
+		return c.socket.Close()
+	}
+	return nil
```

Comment:
https://cs.opensource.google/go/go/+/refs/tags/go1.23.3:src/net/net.go;l=209

Close is not actually sets nil to the socket it uses so you probably expect here to socket to be opened rather than not closed before. In you will pass it in constructor consumer will have to manage this.